sklearn/ensemble/_hist_gradient_boosting/binning.py

```diff
@@ -50,31 +51,62 @@ def _find_binning_thresholds(col_data, max_bins):
     """
     # ignore missing values when computing bin thresholds
     missing_mask = np.isnan(col_data)
-    if missing_mask.any():
+    any_missing = missing_mask.any()
+    if any_missing:
         col_data = col_data[~missing_mask]
+
+    # If sample_weight is not None and 0-weighted values exist, we need to
+    # remove those before calculating the distinct points.
+    if sample_weight is not None:
+        if any_missing:
+            sample_weight = sample_weight[~missing_mask]
+        nnz_sw = sample_weight != 0
+        col_data = col_data[nnz_sw]
+        sample_weight = sample_weight[nnz_sw]
+
     # The data will be sorted anyway in np.unique and again in percentile, so we do it
     # here. Sorting also returns a contiguous array.
-    col_data = np.sort(col_data)
+    sort_idx = np.argsort(col_data)
+    col_data = col_data[sort_idx]
+    if sample_weight is not None:
+        sample_weight = sample_weight[sort_idx]
+
     distinct_values = np.unique(col_data).astype(X_DTYPE)
+
     if len(distinct_values) <= max_bins:
-        midpoints = distinct_values[:-1] + distinct_values[1:]
-        midpoints *= 0.5
+        # Calculate midpoints if distinct values <= max_bins
+        from numpy.lib.stride_tricks import sliding_window_view
```

Comment:
Please move the import statements at the top. There is no reason to use a local import here.